listenbrainz/labs_api/labs/api/recording_from_recording_mbid.py

```diff
@@ -54,45 +56,84 @@ def fetch(self, params, offset=-1, count=-1):
+                                  first_release_date_year AS year
                              FROM recording r
                              JOIN artist_credit ac
                                ON r.artist_credit = ac.id
                              JOIN artist_credit_name acn
                                ON ac.id = acn.artist_credit
                              JOIN artist a
                                ON acn.artist = a.id
+                             JOIN track t
+                               ON t.recording = r.id
+                             JOIN medium m
+                               ON t.medium = m.id
+                             JOIN release rl
+                               ON m.release = rl.id
+                             JOIN release_group rg
+                               ON rl.release_group = rg.id
+                             JOIN release_group_meta rgm
+                               ON rg.id = rgm.id
                             WHERE r.gid
                                IN %s
-                         GROUP BY r.gid, r.id, r.name, r.length, r.comment, ac.id, ac.name
+                         GROUP BY r.gid, r.id, r.name, r.length, r.comment, ac.id, ac.name, rgm.first_release_date_year
                          ORDER BY r.gid'''
 
                 args = [tuple([psycopg2.extensions.adapt(p) for p in mbids])]
-                if count > 0:
-                    query += " LIMIT %s"
-                    args.append(count)
-                if offset >= 0:
-                    query += " OFFSET %s"
-                    args.append(offset)
-
                 curs.execute(query, tuple(args))
 
-                output = []
+                # Build an index of all the fetched recordings and clean up UUID() -> str
+                recording_index = {}
                 while True:
                     row = curs.fetchone()
                     if not row:
                         break
 
                     r = dict(row)
-                    r['recording_mbid'] = str(r['recording_mbid'])
-                    r['[artist_credit_mbids]'] = [str(r) for r in r['artist_credit_mbids']]
+                    mbid = str(r['recording_mbid'])
+                    r['recording_mbid'] = mbid
+                    recording_index[mbid] = dict(row)
+
+                # Finally collate all the results, ensuring that we have one entry with original_recording_mbid for each 
+                # input argument
+                output = []
+                for p in params:
+                    mbid = p['[recording_mbid]'] 
+                    try:
+                        r = copy.copy(recording_index[mbid])
+                    except KeyError:
+                        try:
+                            r = copy.copy(recording_index[redirect_index[mbid]])
+                        except KeyError:
+                            out = {} 
+                            for k in self.outputs():
```

Comment:
I noticed that there's nothing that checks that the fields in `r` contain all of the keys in `self.outputs()`. Is this checked higher up in the hoster?

Reply:
Its not checked, but inherent in the query itself. I'm ok with it as is.